content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+The __fifteenth pass__ does branch tunneling, which makes uninterrupted branch and conditional branch sequences fall through to the last destination in the sequence. Again, correctness needs to be reevaluated given that the number of transitions may change. 
+
+The __sixteenth pass__ linearizes the LTL CFG by replacing graph edges with jumps (unless a jump would be equivalent to falling through to the following instruction). In order to minimize the number of jumps inserted, an OCaml program picks a block ordering that maximizes fall-throughs. A Coq program then goes over the blocks in that order to generate an instruction sequence. Since program transitions may change in this translation, correctness needs to be adjusted here too.
+
+The __seventeenth pass__ removes unused labels. 
+
+The __eighteenth pass__ modifies the linearized code at call and return points to conform to the target architecture's conventions.
+
+The __nineteenth pass__ outputs an AST corresponding to one of the target assembly languages. 
+
+### Getting to Executable Assembly
+
+An OCaml function prints the target assembly AST with proper register and instruction names. This output is then assembled into an object file with an assembler, and linked with other object files (separately compiled or in the same pass). `gcc` is the most common external assembler and linker used to do this. These three passes are unverified. 
+
+# CompCert Correctness Claims
+
+CompCert promises that it will not mistranslate code it compiles. Specifically, the executable it generates will not perform any action that the source code could not perform. 
+
+## Semantic Preservation, Intuitively
+
+*Semantic preservation* requires that the compiled code's behavior (generally, the trace of all I/O and mutable operations as well as the exit code) is a refinement (or subset) of the source code's behavior, where refinement may include the removal of diverging dead code, as in the twelfth pass. What does this look like formally? 
+
+## Specifying Preservation with Simulations 
+
+CompCert has a semantics of source and target language code behavior. Given such a semantics, we can formulate different notions of *simulation* such that two programs are equal if they are related by simulation. 
+
+Let a source language program $\mathbf{s}$ begins at state $s_0$, runs for $n$ steps, and ends at state $s_n$. Let a target language program $\mathbf{t}$ begins at state $t_0$, runs for $m$ steps, and ends at state $t_m$. State transitions may have effects, so let each transition $s_i \rightarrow s_j$ be accompanied by a label $e_{s_j}$ specifying the potential effect.
+
+### Bisimulation
+
+A *bisimulation* relation between the source and target programs is a relation $R$ with the following conditions: 
```

Comment:
Maybe worth one sentence summarizing the definition informally (i.e., without math)?